src/nvim/terminal.c

```diff
@@ -323,6 +323,7 @@ void terminal_close(Terminal *term, char *msg)
       term->opts.close_cb(term->opts.data);
     }
   } else {
+    refresh_terminal(term);
```

Comment:
If `terminal_close()` was called with `msg != NULL` and `exiting == TRUE`, `on_exit` could still be called with a stale terminal buffer. So it could be safer to move `refresh_terminal(term);` in front of this `if`-block.